internal/configs/provider.go

```diff
@@ -23,7 +25,12 @@ type Provider struct {
 	Name       string
 	NameRange  hcl.Range
 	Alias      string
-	AliasRange *hcl.Range // nil if no alias set
+	AliasExpr  hcl.Expression // nil if no alias set
+	AliasRange *hcl.Range     // nil if no alias set
+	ForEach    hcl.Expression
+	EachValue  *cty.Value
+	Count      hcl.Expression
+	CountIndex *cty.Value
```

Comment:
Two things that quite bother me with this implementation:
1. This struct is no longer as described in the comment above: `Provider represents a "provider" block in a module or file.` It sometimes represents a block and sometimes a provider "instance".
2. Now, it holds two responsibilities and changes based on the application's lifecycle. If `tofu` is in the stage of reading the configuration, this struct represents the `provider` block. If `tofu` is building the module tree and populating the providers (after calling `decodeStaticFields()`), the struct now describes a specific "instance" of the provider.

This implementation is quite confusing and different from the practices in the tofu repository. Maybe we should consider adding a different struct that will include the `EachValue` and `CountIndex`?

Reply:
1. It's more so a "provider block" vs "provider block with instance data"
2. This is true, the expansion here feels a bit weird IMO and I noticed that during the prototype phase.  This change is the "low churn" option, but adds cognitive load to understanding provider configs in the configs package. 

There are technically 3 levels of config parsing that are all mixed together:
1. hcl/json text -> configs.File (HCL loading)
2. configs.File -> configs.Module (merging files)
3. configs.Module -> configs.Module (static evaluation)

Even before static evaluation was added, there were all sorts of odd validation bugs all throughout the code when dealing with overrides.  Static evaluation shines more of a light on this.

In this specific scenario, we would need:
```go

// File parsing produces
type configs.ProviderBlock struct {
  // 1-1 mapping to HCL
}

// Module struct from Files (normal+override) produces
type configs.ProviderData struct {
  // Closer to existing struct
  ForEach hcl.Expression
  Count hcl.Expression
}
// Validation is performed when constructing ProviderData

// func (ProviderData) expand(StaticContext) []Provider

// Expanded into one or more
type configs.Provider struct {
  // Matches existing struct
  
  ForEachValue *cty.Value
  CountValue *cty.Value
  // OR
  InstanceData instances.RepetitionData
}
```


For the purposes of this PR, we could merge the concept of ProviderBlock and ProviderData to reduce the potential set of changes as it is a less important distinction IMO.

Reply:
I agree ^
Also, I prefer the usage of `InstanceData instances.RepetitionData` over `ForEachValue` and `CountValue`.

Reply:
I agree ^
Also, I prefer the usage of InstanceData instances.RepetitionData over ForEachValue and CountValue.

Reply:
We already track some of the modernization/refactoring issues and I would suggest we document that down to address later. Both @Evi1Pumpkin and @cam72cam arguments sound pretty reasonable.

Breaking down config package structures into different stages such as loading (raw) and evaluation (ready for graph processing) makes perfect sense to me.

Reply:
We discussed this offline. The agreement was to go with @cam72cam 's suggestion and do 2 structs: one that captures the understanding of "pre-evaluation" and one that is "post evaluation". Then we will revisit in future with an RFC on larger refactoring work.